test/e2e_env/kubernetes/membership/membership.go

```diff
@@ -0,0 +1,86 @@
+package membership
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/kumahq/kuma/test/e2e_env/kubernetes/env"
+	. "github.com/kumahq/kuma/test/framework"
+	"github.com/kumahq/kuma/test/framework/deployments/testserver"
+)
+
+func Membership() {
+	const ns1 = "membership-1"
+	const ns2 = "membership-2"
+	const mesh1 = "membership-1"
+	const mesh2 = "membership-2"
+
+	meshAllowingNamespace := func(mesh, ns string) string {
+		return fmt.Sprintf(`apiVersion: kuma.io/v1alpha1
+kind: Mesh
+metadata:
+  name: %s
+spec:
+  constraints:
+    dataplaneProxy:
+      requirements:
+      - tags:
+          k8s.kuma.io/namespace: %s`, mesh, ns)
+	}
+
+	It("should take into account membership when dp is connecting to the CP", func() {
+		E2EDeferCleanup(func() {
+			Expect(env.Cluster.TriggerDeleteNamespace(ns1)).To(Succeed())
+			Expect(env.Cluster.DeleteMesh(mesh1)).To(Succeed())
+			Expect(env.Cluster.TriggerDeleteNamespace(ns2)).To(Succeed())
+			Expect(env.Cluster.DeleteMesh(mesh2)).To(Succeed())
+		})
+		// given
+		Expect(NamespaceWithSidecarInjection(ns1)(env.Cluster)).To(Succeed())
+		Expect(YamlK8s(meshAllowingNamespace(mesh1, ns1))(env.Cluster)).To(Succeed())
+		Expect(NamespaceWithSidecarInjection(ns2)(env.Cluster)).To(Succeed())
+		Expect(YamlK8s(meshAllowingNamespace(mesh2, ns2))(env.Cluster)).To(Succeed())
```

Comment:
nit: I like to chain such things with `NewClusterSetup()`, but that's only my preference.